### PR TITLE
Adds CBOR ByteString serialization logic

### PR DESCRIPTION
**Pull request by Chris-Corea**
# What is this?
This change allows for the encoding of fixed-length `ByteStrings` and the decoding of definite-length or indefinite-length `ByteStrings`.

Encoding and Decoding "rules" can be found in [Section 2.1 of the CBOR RFC](https://tools.ietf.org/html/rfc7049#section-2.1) (major type 2). 

Information about treating indefinite-length byte strings can be found in [Section 2.2.2 of the CBOR RFC](https://tools.ietf.org/html/rfc7049#section-2.2.2).

## Does it work?
Sure does! But this implementation still requires the consumer to provide their own data serializer. That is, `ByteArray`s will still be serialized as a list of bytes (as is the current default behavior), rather than byte strings. ~I'll noodle around with the code to see if there's a way we could get around that.~ _see update_

Example class with a simple serializer:
```kotlin
@Serializable
data class BinaryPayload(
    val data: ByteArray
) {
    @Serializer(forClass = BinaryPayload::class)
    companion object : KSerializer<BinaryPayload> {
        override val descriptor: SerialDescriptor = SerialDescriptor(
            serialName = "cbor.test.BinaryPayload", 
            kind = StructureKind.MAP
        ) {
            element<ByteArray>("data")
        }

        override fun serialize(encoder: Encoder, value: BinaryPayload) {
            val compEncoder = encoder.beginStructure(descriptor)
            compEncoder.encodeStringElement(descriptor, 0, "data")
            compEncoder.encodeByteStringElement(descriptor, 1, value.data)
            compEncoder.endStructure(descriptor)
        }

        override fun deserialize(decoder: Decoder): BinaryPayload {
            val compDecoder = decoder.beginStructure(descriptor)
            var bytes: ByteArray? = null
            loop@ while (true) {
                when (val i = compDecoder.decodeElementIndex(descriptor)) {
                    CompositeDecoder.READ_DONE -> break@loop
                    0 -> compDecoder.decodeStringElement(descriptor, i) // the key in the CBOR map is unimportant in this example. just move the buffer along.
                    1 -> bytes = compDecoder.decodeByteStringElement(descriptor, i)
                    else -> throw SerializationException("Unknown index: $i")
                }
            }
            compDecoder.endStructure(descriptor)

            return BinaryPayload(
                bytes ?: throw MissingFieldException("data")
            )
        }
    }
}
```

# Update!!

I was able to add a new type to `runtime/commonMain` called `ByteString`, which includes its own `Serializer`. This means that the above code can turn into:
```kotlin
@Serializable
data class BinaryPayload(
    val data: ByteString
)
```

### CODE REVIEW

File: runtime/commonMain/src/kotlinx/serialization/internal/ByteString.kt [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+expect class ByteString(bytes: ByteArray) {
+    val bytes: ByteArray
+    val size: Int
+
+    @Serializer(forClass = ByteString::class)
+    companion object : KSerializer<ByteString> {
+        override val descriptor: SerialDescriptor
+        override fun serialize(encoder: Encoder, value: ByteString)
+        override fun deserialize(decoder: Decoder): ByteString
+    }
```
Comment (Chris-Corea) [inline runtime/commonMain/src/kotlinx/serialization/internal/ByteString.kt:19]: Unfortunately for us, the `Serializer` look up for each object/class happens at compile time. It's actually great to have compile time checks like that, however it makes it so that any new type to be supported by `kotlinx.serialization` needs to be defined in a [look-up table in the Kotlin compiler](https://github.com/JetBrains/kotlin/blob/379c6944a20f17e02f0ce9827adb6683b99f672d/plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/common/TypeUtil.kt#L169-L213). This works around not being able to (for good reason) add to the list of types defined in the compiler. 

The downside is that each platform's implementation of `ByteString` will need to provide their own implementation of this serializer. It's not the end of the world, but a bit of a code smell that I'm not super happy about.

File: formats/cbor/commonTest/src/kotlinx/serialization/cbor/SampleClasses.kt [hunk 2]
```diff
@@ -5,27 +5,34 @@
...
-        val map: Map<Int, Boolean>,
-        val inner: Simple,
-        val innersList: List<Simple>
+    val str: String,
+    val i: Int,
+    val nullable: Double?,
+    val list: List<String>,
+    val map: Map<Int, Boolean>,
+    val inner: Simple,
+    val innersList: List<Simple>
```
Comment (twyatt) [inline formats/cbor/commonTest/src/kotlinx/serialization/cbor/SampleClasses.kt:21]: Despite seemingly incorrect, I wouldn't adjust indentation of existing code (if you don't otherwise have to modify it).

IMHO, if you _add_ code, then it makes sense to use the "correct" indentation (if you prefer), but keep PRs as small as possible.

File: formats/cbor/commonTest/src/kotlinx/serialization/cbor/SampleClasses.kt [hunk 3]
```diff
@@ -5,27 +5,34 @@
...
-        val float: Float,
-        val double: Double,
-        val boolean: Boolean,
-        val char: Char
+    val int: Int,
+    val long: Long,
+    val float: Float,
+    val double: Double,
+    val boolean: Boolean,
+    val char: Char
```
Comment (twyatt) [inline formats/cbor/commonTest/src/kotlinx/serialization/cbor/SampleClasses.kt:31]: Revert indentation changes, please.

**Comment by Chris-Corea** [comment]
## Incoming wall of text

Thanks for the review @twyatt! I haven't gone through all of your comments but wanted to address your first two questions:

> Looks like some guidelines for these types of additions were outlined in Kotlin#52 (comment). Are we following the recommended design?

I hadn't so thanks for sending the link. Luckily, it seems like what I did follows the recommended design. The only "gray area" that I think could be problematic is the introduction of a new, built-in type. But I think that's what makes the `ByteString` implementation a bit of a challenge, since CBOR allows for raw bytes in the data payload, whereas JSON or Protobuf require an object structure or an array. Which leads me to...

> Additionally, looks like protobuf got ByteArray support, does it offer any insight into a good direction for how to implement ByteArray (encoded as ByteString) for CBOR? Kotlin#509

What's especially nice about the project structure/design is that the changes made in [Kotlin#509](https://github.com/Kotlin/kotlinx.serialization/pull/509) to add serializers for arrays of primitive types carry over to the JSON and CBOR formats. My initial idea for adding `ByteString` support involved copying and repurposing the `ByteArraySerializer`, but that doesn't work because

1. The primitive-type array serializers (including the `ByteArraySerializer`) extend `ListLikeDescriptor` which has a `serialKind` of `StructureKind.LIST`. Sure, I could write my own serializer that extends `ByteArraySerializer` (seems most relevant) and override the relevant properties and methods, but at that point I've basically completely changed the logic defined by `ListLikeDescriptor`, which, to me, is a red flag that I shouldn't be extending the class at all. Also, this approach doesn't solve for the differences in encoding an array vs a ByteString. And,

2. In CBOR, a `ByteString` is much more like a `String` than an array, so treating its serialization as though it were an array overcomplicates the implementation. There's no need to "serialize" individual bytes one-by-one since we can just write the entire array of bytes to the output stream. 

There's no need for marking the begin or end of the encoded type like you would for a map or an array. A `ByteString` (and regular CBOR string) just has a header (which, arguably, could be thought of as the "begin" marker) that indicates the string's size. 

So now maybe you're wondering, "well, then why not use the built-in `StringSerializer`?" It's because the type expected as input by the `StringSerializer` is a _`String`_ (which is maybe obvious). `Serializers` make use of functions defined in the `Encoding` and `Decoding` interfaces that are written per primitive type (i.e. `encodeString(...)`, `encodeByte(...)`, `decodeByte()`). My takeaway: 

* define functions in the `Encoding` and `Decoding` interfaces to be invoked by the serializer and implemented by the format encoder/decoder. _Just this alone adds support for encoding and decoding ByteStrings. The downside is that the consumer has to provide their own serializer for any object that needs to be serialized as a ByteString._

* Extend the existing `StringSerializer` which isn't possible because it's an `object` so...
 
* Create a new string-like serializer that accepts `ByteArray`s as input and outputs a `ByteArray`. This is good, but still not great since the consumer would still need to explicitly call out the serializer for _each_ `ByteArray` property in the object to serialize. This actually introduces another issue because I _think_ providing a custom serializer for a built-in type would override the built-in serializer. So if a consumer wanted to make use of serializing a `ByteArray` as an array of bytes _and_ a ByteString, this approach would prevent them from being able to do so. Example:
```kotlin
@Serializable
data class Signature(
    @Serializable(with = CustomSerializer::class) val data: ByteArray
)

@Serializer(forClass = ByteArray)
object CustomSerializer : KSerializer<ByteArray> { ... }
```

* Create a new type "ByteString" with a built-in serializer. The type is backed by a ByteArray and allows the consumer to make use of the built-in ByteArray serialization logic as well as the new, ByteString specific serialization logic. And what's nice about this is that now we can add the new type and its serializer to the built-in map of known serializers. So the consumer's code goes from the above to:
```kotlin
@Serializable
data class Signature(
    val data: ByteString
)
```

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt [hunk 4]
```diff
@@ -212,6 +221,7 @@ public class Cbor(
         override fun decodeBoolean() = decoder.nextBoolean()
 
         override fun decodeByte() = decoder.nextNumber().toByte()
+        override fun decodeByteString(): ByteString = ByteString(decoder.nextByteString())
```
[the remark this answers is not in the document]
Comment (Chris-Corea) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:224]: `ByteArray` serialization is handled by encoding or decoding each individual `Byte` in the collection one by one. The `ByteArraySerializer` extends the `PrimitiveArraySerializer` which encodes and decodes a "start" marker before the first element and an "end" marker after the last element to create the array structure.

Whereas with `ByteString`, we just need to write the string header to indicate the type and size of the data, and then simply write the bytes to the output stream. They could probably do something similar with ByteArray.

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt [hunk 5]
```diff
@@ -294,6 +305,33 @@ public class Cbor(
             return res
         }
 
+        fun nextByteString(): ByteArray {
```
[the remark this answers is not in the document]
Comment (Chris-Corea) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:308]: I'm leaning towards leaving it as is, but this is maybe something we should talk about since I call this function recursively when parsing an indefinite-length `ByteString`. In the `CborReader` class (just above this comment), the decoder uses this function like:

```kotlin
override fun decodeByteString() = ByteString(decoder.nextByteString())
```

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt [hunk 6]
```diff
@@ -294,6 +305,33 @@ public class Cbor(
...
+            val bytes = when (bytesToRead) {
+                0 -> input.readExactNBytes(additionalData)
+                -1 -> {
+                    readByte()
+                    readIndefiniteLengthByteString()
+                }
+                else -> input.readBytes(input.readExact(bytesToRead))
+            }
+
+            readByte()
```
[the remark this answers is not in the document]
Comment (Chris-Corea) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:331]: It moves the reader's place over to the start of the next CBOR object. I don't like it either, but it's what they do for the other object types.

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt [hunk 7]
```diff
@@ -322,6 +360,20 @@ public class Cbor(
...
+         */
+        private fun InputStream.readBytes(length: Long): ByteArray {
+            if (length > Int.MAX_VALUE) {
+                val firstChunk = input.readExactNBytes(length.toInt()) // least signficant 32 bits
+                val secondChunk = input.readExactNBytes((length shr 32).toInt()) // remaining value
+                return firstChunk + secondChunk
+            } else {
+                return input.readExactNBytes(length.toInt())
+            }
+        }
```
[the remark this answers is not in the document]
Comment (Chris-Corea) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:375]: Wow... That is unfortunate. You're right, that this was intended to get around the limit imposed by the `Int` parameter in `readExactNBytes`, but also to return a `ByteArray` rather than a `Long`.

I'll try to look into Jackson's implementation as to how they handle indefinite-length `ByteString`s.

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt [hunk 8]
```diff
@@ -353,6 +405,32 @@ public class Cbor(
...
+        /** Indefinite-length byte strings contain an unknown number of definite-length byte strings (chunks).
+         * Chunks should always provide their own lengths, that is, nested, indefinite-length
+         * byte strings are not allowed per the CBOR RFC.
+         *
+         * @return A [ByteArray] containing all of the concatenated byte strings found in the buffer.
+         * @throws [CborDecodingException] if invoked while already reading an indefinite-length byte string.
+         */
+        private fun readIndefiniteLengthByteString(): ByteArray {
+            if (isReadingUnknownLengthByteString)
+                throw CborDecodingException("Nested indefinite-length bytestrings are not allowed", curByte)
```
[the remark this answers is not in the document]
Comment (Chris-Corea) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:418]: Right, the goal is to ensure we conform to the RFC.

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt [hunk 9]
```diff
@@ -353,6 +405,32 @@ public class Cbor(
...
+                val bytes = nextByteString()
+                buffer = if (buffer == null) bytes else buffer + bytes
+            } while (!isEnd())
+            isReadingUnknownLengthByteString = false
+
+            // advance to the next byte
+            end()
+
+            return buffer
+                ?: throw CborDecodingException("Unable to decode chunk of indefinite length ByteString.", curByte)
```
[the remark this answers is not in the document]
Comment (Chris-Corea) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:432]: They are allowed. Good catch.

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (Chris-Corea) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:375]: I remembered the main reason why I needed this function. It's because the additional data could specify a byte string length of 8 bytes. And if that were the case, I'd need two `ByteArray`s to hold the data. :(

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (twyatt) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:331]: Since we know the value of the "end" byte, we should probably validate it here.

Something like:

```kotlin
val endByte = readByte()
check(endByte == END_BYTE) { "Unexpected byte for CBOR end: $endByte" }
```

_Better names could be chosen for the variables. i.e. try to match CBOR RFC naming if possible._

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (twyatt) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:375]: I don't quite follow why you'd need 2 `ByteArray`s?

[the remark this answers is not in the document]
Comment (Chris-Corea) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:375]: If we ever encounter a byte string who's size is greater than 4 bytes, we'd need two `ByteArrays` to hold the information. The CBOR RFC allows for a case where a byte string's _length_ is a 64 bit value. 

That means for any fixed-length `ByteString`, we need to account for  a string who's length is at most `Long.MAX_VALUE`. We discussed on Zoom that it's pretty crazy for a string that length to be sent over a protocol making use of CBOR, but the RFC allows it.

So I think I need:

1. to keep this function (allows for reading an 8-byte length;
2. add another `ByteArray` to the `ByteString` class to hold the second half of the 8-byte length ByteString (which really doesn't solve for the case where the byte string's length is `Long.MAX_VALUE`, considering that the maximum "safe" size limit on arrays is `Int.MAX_VALUE - 8`. Do I need 3 `ByteArrays` to be safe?);
3. add logic to `ByteString` to support indexing over two separate arrays (i.e. `byteString[Int.MAX_VALUE+1] // retrieves the first element of the second `ByteArray`).

Let me know if you want to jump on Zoom again. Either way, I think I might need to still look at how Jackson handles this particular case.

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (Chris-Corea) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:335]: Ah, so to clarify: we'll only see the end byte (`0xFF`) at the end of an indefinite-length bytestring.

Both of the below examples are valid CBOR.
```
5F // indefinite-length bytestring: major type two, additional data 31
  4C // fixed-length bytestring of size 12 
    6332383500000000000001AD // 12 bytes
FF // end of indefinite-length bytestring

OR

4C // fixed-length bytestring of size 12
  6332383500000000000001AD // 12 bytes of data
63 // utf-8 text of size 3
  626172 // 3 bytes of data
```

The `readIndefiniteLengthByteString` function's `while` loop is set up to break out of the loop once the end marker (`0xFF`) is encountered.

[the remark this answers is not in the document]
Comment (twyatt) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:335]: Ok, maybe I now understand what you meant by the library being a little weird, in that you have to advance the cursor after you're done reading? Is that what `readByte` is doing (essentially allowing later reads to have the read cursor **after** the CBOR data we were reading)?

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt [hunk 10]
```diff
@@ -367,12 +367,12 @@ public class Cbor(
...
          * containing the read bytes.
+         *
+         * @throws [SerializationException] if the provided length does not fit in a 32-bit signed [Int].
          */
         private fun InputStream.readBytes(length: Long): ByteArray =
             if (length > Int.MAX_VALUE) {
-                val firstChunk = input.readExactNBytes(length.toInt()) // least signficant 32 bits
-                val secondChunk = input.readExactNBytes((length shr 32).toInt()) // remaining value
-                firstChunk + secondChunk
+                throw SerializationException("Unsupported length for current byte: $curByte")
```
Comment (twyatt) [inline formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:375]:
```suggestion
                throw SerializationException("Unsupported length $length for current byte: $curByte")
```

File: runtime/commonMain/src/kotlinx/serialization/Encoding.kt [hunk 11]
```diff
@@ -140,6 +141,13 @@ public interface Encoder {
      */
     public fun encodeByte(value: Byte)
 
+    /**
+     * Encodes an array of bytes as a CBOR ByteString.
+     *
+     * Can be ignored by all other formats.
+     */
+    public fun encodeByteString(value: ByteString) {}
```
Comment (twyatt) [inline runtime/commonMain/src/kotlinx/serialization/Encoding.kt:149]: Rather than an empty function, can we throw `UnsupportedOperation` (or similar) except for CBOR encoding?

File: runtime/commonMain/src/kotlinx/serialization/Encoding.kt [hunk 12]
```diff
@@ -322,6 +330,12 @@ public interface CompositeEncoder {
      */
     public fun encodeByteElement(descriptor: SerialDescriptor, index: Int, value: Byte)
 
+    /**
+     * Encodes the bytes [value] associated with an element at the given [index] in [serial descriptor][descriptor].
+     * The element at the given [index] should be a [ByteArray].
+     */
+    public fun encodeByteStringElement(descriptor: SerialDescriptor, index: Int, value: ByteString) {}
```
Comment (twyatt) [inline runtime/commonMain/src/kotlinx/serialization/Encoding.kt:337]: Rather than an empty function, can we throw `UnsupportedOperation` (or similar) except for CBOR encoding?

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (twyatt) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt]: We landed on: Jackson doesn't support it so neither will we, right?
So option 4, just throw an exception.

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (twyatt) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:421]: Hmm, I can't think of a better way, but it worries me to have the mutable state.
If this method is called from multiple threads then this would sporadically not work as expected.
Perhaps the serialization is single threaded and we're safe here. ¯\\_(ツ)\_/¯

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt [hunk 13]
```diff
@@ -353,6 +408,40 @@ public class Cbor(
...
+            val output = ByteArray(sumBy { it.size })
+
+            var position = 0
+            for (chunk in this) {
+                chunk.copyInto(output, position)
+                position += chunk.size
+            }
+
+            return output
+        }
```
Comment (twyatt) [inline formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:444]: Move to outside the class. It enforces that function is pure (being outside the class it won't have access to any state in the class).

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (Chris-Corea) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt]: Yup! That's right.

File: runtime/commonMain/src/kotlinx/serialization/Encoding.kt (hunk 11 above)
Comment (Chris-Corea) [reply runtime/commonMain/src/kotlinx/serialization/Encoding.kt:149]: Yeah, good idea. I'll update.

File: runtime/commonMain/src/kotlinx/serialization/Encoding.kt [hunk 14]
```diff
@@ -140,6 +141,13 @@ public interface Encoder {
      */
     public fun encodeByte(value: Byte)
 
+    /**
+     * Encodes an array of bytes as a CBOR ByteString.
+     *
+     * Can be ignored by all other formats.
+     */
+    public fun encodeByteString(value: ByteString) { throw SerializationException("") }
```
Comment (twyatt) [inline runtime/commonMain/src/kotlinx/serialization/Encoding.kt:149]: Hmm, unfortunate they didn't give you a constructor w/o a `message`.
Since it's required though, using `""` feels like an abuse.
Is there some small message you can put in there to give developer hint of what went wrong when this exception is thrown?

File: runtime/commonMain/src/kotlinx/serialization/Encoding.kt [hunk 15]
```diff
@@ -322,6 +330,12 @@ public interface CompositeEncoder {
      */
     public fun encodeByteElement(descriptor: SerialDescriptor, index: Int, value: Byte)
 
+    /**
+     * Encodes the bytes [value] associated with an element at the given [index] in [serial descriptor][descriptor].
+     * The element at the given [index] should be a [ByteArray].
+     */
+    public fun encodeByteStringElement(descriptor: SerialDescriptor, index: Int, value: ByteString) { throw SerializationException("") }
```
Comment (twyatt) [inline runtime/commonMain/src/kotlinx/serialization/Encoding.kt:337]: Same here.

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt [hunk 16]
```diff
@@ -294,6 +310,32 @@ public class Cbor(
             return res
         }
 
+        fun nextByteString(): ByteArray {
+            if ((curByte and 0b111_00000) != HEADER_BYTESTRING.toInt())
+                throw CborDecodingException("start of a ByteString.", curByte)
+
+            val additionalData = curByte and 0b000_11111
+            val bytesToRead: Long = when (additionalData) {
```
Comment (twyatt) [inline formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:318]:
```suggestion
            val bytesToRead: Long = when (val additionalData = curByte and 0b000_11111) {
```

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt [hunk 17]
```diff
@@ -322,6 +364,19 @@ public class Cbor(
...
+         * Reads the specified number of bytes from the input stream and returns a [ByteArray]
+         * containing the read bytes.
+         *
+         * @throws [SerializationException] if the provided length does not fit in a 32-bit signed [Int].
+         */
+        private fun InputStream.readBytes(length: Long): ByteArray =
+            if (length > Int.MAX_VALUE) {
+                throw SerializationException("Unsupported length: $length, for current byte: $curByte")
+            } else {
+                input.readExactNBytes(length.toInt())
```
Comment (twyatt) [inline formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:377]:
```suggestion
                readExactNBytes(length.toInt())
```

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt [hunk 18]
```diff
@@ -322,6 +364,19 @@ public class Cbor(
             return result
         }
 
+        /**
+         * Reads the specified number of bytes from the input stream and returns a [ByteArray]
+         * containing the read bytes.
+         *
+         * @throws [SerializationException] if the provided length does not fit in a 32-bit signed [Int].
+         */
+        private fun InputStream.readBytes(length: Long): ByteArray =
```
Comment (twyatt) [inline formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:373]: Please move outside of class to enforce that it's a "pure function".

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt [hunk 19]
```diff
@@ -322,6 +364,19 @@ public class Cbor(
             return result
         }
 
+        /**
+         * Reads the specified number of bytes from the input stream and returns a [ByteArray]
+         * containing the read bytes.
+         *
+         * @throws [SerializationException] if the provided length does not fit in a 32-bit signed [Int].
```
Comment (twyatt) [inline formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:371]: Brackets not needed after `@throws`.

```suggestion
         * @throws SerializationException if the provided length does not fit in a 32-bit signed [Int].
```

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt [hunk 20]
```diff
@@ -353,6 +408,28 @@ public class Cbor(
...
+            if (isReadingUnknownLengthByteString)
+                throw CborDecodingException("Nested indefinite-length bytestrings are not allowed", curByte)
+
+            val byteStrings = mutableListOf<ByteArray>()
+
+            isReadingUnknownLengthByteString = true
+            do {
+                byteStrings.add(nextByteString())
+            } while (!isEnd())
+            isReadingUnknownLengthByteString = false
```
Comment (twyatt) [inline formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:429]: A little safer as:

```suggestion
            try {
                isReadingUnknownLengthByteString = true
                do {
                    byteStrings.add(nextByteString())
                } while (!isEnd())
            } finally {
                isReadingUnknownLengthByteString = false
            }
```

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt [hunk 21]
```diff
@@ -212,6 +226,7 @@ public class Cbor(
         override fun decodeBoolean() = decoder.nextBoolean()
 
         override fun decodeByte() = decoder.nextNumber().toByte()
+        override fun decodeByteString(): ByteString = ByteString(decoder.nextByteString())
```
Comment (twyatt) [inline formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:229]: nitpick: Return type is redundant here. More so, we should follow convention of the functions near it.

```suggestion
        override fun decodeByteString() = ByteString(decoder.nextByteString())
```

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt [hunk 22]
```diff
@@ -12,6 +12,12 @@ import kotlinx.serialization.internal.*
 import kotlinx.serialization.modules.*
 import kotlin.experimental.*
 
+private const val ADD_DATA_ONE_BYTE = 24
+private const val ADD_DATA_TWO_BYTES = 25
+private const val ADD_DATA_FOUR_BYTES = 26
+private const val ADD_DATA_EIGHT_BYTES = 27
+private const val ADD_DATA_INDEF_LENGTH = 31
```
Comment (twyatt) [inline formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:19]: Not sure how these are defined in the RFC, so feel free to disregard my suggestion if it doesn't make sense here (i.e. if RFC lists as numbers, not binary, then disregard).

```suggestion
private const val ADD_DATA_ONE_BYTE = 0b0001_1000 // 24
private const val ADD_DATA_TWO_BYTES = 0b0001_1001 // 25
private const val ADD_DATA_FOUR_BYTES = 0b0001_1010 // 26
private const val ADD_DATA_EIGHT_BYTES = 0b0001_1011 // 27
private const val ADD_DATA_INDEF_LENGTH = 0b0001_1111 // 31
```

File: runtime/commonMain/src/kotlinx/serialization/Encoding.kt (hunk 14 above)
Comment (Chris-Corea) [reply runtime/commonMain/src/kotlinx/serialization/Encoding.kt:149]: Yeah, definitely, I'll add a message. I should have done this from the start.

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt (hunk 16 above)
Comment (Chris-Corea) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:318]: AS suggested this, but I feel like it isn't as easily read. Is this a common pattern?

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt (hunk 17 above)
Comment (Chris-Corea) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:377]: 🤦‍♂️

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt (hunk 18 above)
Comment (Chris-Corea) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:373]: I kept this inside the class since that's where they had written their `InputStream` extension. I agree with you that this should move to outside of the class, but it's more in-line with what they've written as it is.

Maybe I'll move it outside the class and if they really think it's better to have it defined _inside_ the class, then wait until they suggest it?

Comment (Chris-Corea) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:373]: Moving this outside of the class also means that the exception message will need to drop the use of `curByte`, which isn't a big deal, but just wanted to make sure you're aware.

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt (hunk 22 above)
Comment (Chris-Corea) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:19]: RFC does list these values in the decimal format. Good idea though.

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt (hunk 16 above)
Comment (twyatt) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:318]: It is a common pattern, yes.

The advantage of putting the `additionalData` within the `when` is that the variable is scoped to the `when` block.

I think it's fine if you feel it's more readable outside of the `when`, I'll leave it up to you.

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt (hunk 18 above)
Comment (twyatt) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt]: Oh, I missed that it was using the `curByte` property.

I personally like it when exceptions give me as many hints as possible as to what went wrong. So I think it's worth keeping in the class definition.

**tl;dr** disregard my original comment.

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt (hunk 22 above)
Comment (twyatt) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:19]: I just saw that they used the binary representation elsewhere in this class, but if the RFC lists them as numbers, then I think it's best to stick with the "documentation" (the RFC) for others that may read the code.

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt (hunk 16 above)
Comment (Chris-Corea) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:318]: > The advantage of putting the additionalData within the when is that the variable will be scoped to the when block.

Oh, cool. I didn't know that. I'll update. 👍

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (twyatt) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:420]: Being that `curByte` is relied upon indicates that they designed around mutable state being held during the serialization process, I think we're good. 👍
